pkg/controller/networkpolicy/validate_test.go

```diff
@@ -1528,14 +1616,220 @@ func TestValidateAntreaGroup(t *testing.T) {
 		},
+		{
+			name: "anp-group-set-with-podselector-specified-to-update",
+			currGroup: &crdv1alpha3.Group{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "anp-group-set-with-podselector-specified-to-update",
+					Namespace: "x",
+				},
+				Spec: crdv1alpha3.GroupSpec{
+					PodSelector: &metav1.LabelSelector{
+						MatchLabels: map[string]string{"foo=": "bar"},
+					},
+				},
+			},
+			oldGroup: &crdv1alpha3.Group{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "anp-group-set-with-podselector-and-ipblock",
+					Namespace: "x",
+				},
+				Spec: crdv1alpha3.GroupSpec{
+					PodSelector: &metav1.LabelSelector{
+						MatchLabels: map[string]string{"foo=": "bar"},
+					},
+					IPBlocks: []crdv1alpha1.IPBlock{
+						{CIDR: "10.0.0.10/32"},
+					},
+				},
+			},
+			operation: admv1.Update,
+		},
+		{
+			name: "to-delete",
+			oldGroup: &crdv1alpha3.Group{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "anp-group-set-with-podselector-specified-to-update",
+					Namespace: "x",
+				},
+				Spec: crdv1alpha3.GroupSpec{
+					PodSelector: &metav1.LabelSelector{
+						MatchLabels: map[string]string{"foo=": "bar"},
+					},
+				},
+			},
+			operation: admv1.Delete,
+		},
 	}
 	for _, tt := range tests {
 		t.Run(tt.name, func(t *testing.T) {
-			_, c := newController()
-			v := NewNetworkPolicyValidator(c.NetworkPolicyController)
-			actualReason, allowed := v.validateAntreaGroup(tt.group, nil, admv1.Create, authenticationv1.UserInfo{})
+			_, controller := newController()
+			validator := NewNetworkPolicyValidator(controller.NetworkPolicyController)
+			actualReason, allowed := validator.validateAntreaGroup(tt.currGroup, tt.oldGroup, tt.operation, authenticationv1.UserInfo{})
+			assert.Equal(t, tt.expectedReason, actualReason)
+			if tt.expectedReason == "" {
+				assert.True(t, allowed)
+			} else {
+				assert.False(t, allowed)
+			}
+		})
+	}
+}
+
+func TestValidateTier(t *testing.T) {
+	tests := []struct {
+		name      string
+		currTier  *crdv1alpha1.Tier
+		oldTier   *crdv1alpha1.Tier
+		operation admv1.Operation
+	}{
+		{
+			name: "create-validate-tier",
+			currTier: &crdv1alpha1.Tier{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "tier-priority-3",
+					Namespace: "x",
+				},
+				Spec: crdv1alpha1.TierSpec{
+					Priority: 3,
+				},
+			},
+			operation: admv1.Create,
+		},
+		{
+			name: "delete-tier",
+			oldTier: &crdv1alpha1.Tier{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "tier-priority-3",
+					Namespace: "x",
+				},
+				Spec: crdv1alpha1.TierSpec{
+					Priority: 3,
+				},
+			},
+			operation: admv1.Delete,
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			_, controller := newController()
+			validator := NewNetworkPolicyValidator(controller.NetworkPolicyController)
+			actualReason, allowed := validator.validateTier(tt.currTier, tt.oldTier, tt.operation, authenticationv1.UserInfo{})
+			if actualReason == "" {
+				assert.True(t, allowed)
```

Comment:
The check doesn't make sense here, it will always be allowed when actualReason is empty, you should provide an `expectedReason` field in the test structure like other test functions. please add cases to cover failed validation cases.